crates/types/src/epoch_table.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2024 - Restate Software, Inc., Restate GmbH.
+// All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+#![allow(dead_code)]
+
+use crate::identifiers::{LeaderEpoch, PartitionId};
+use crate::{GenerationalNodeId, Version, Versioned};
+use std::collections::HashMap;
+
+/// Partition processor epoch table.
+#[derive(Debug, Clone, Default)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EpochTable {
+    version: Version,
+    epochs: HashMap<PartitionId, EpochMetadata>,
+}
+
+#[derive(Debug, Clone)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub struct EpochMetadata {
+    leader_epoch: LeaderEpoch,
+    /// Node on which the leading partition processor should run
+    node_id: MaybeNodeId,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+enum MaybeNodeId {
+    Assigned(GenerationalNodeId),
+    Unassigned,
+}
+
+impl EpochTable {
+    pub fn increment_version(&mut self) {
+        self.version = self.version.next();
+    }
+
+    pub fn get(&self, partition_id: &PartitionId) -> Option<&EpochMetadata> {
```

Comment:
nit, if PartitionId is Copy (and I think it should be), then better pass by value (the compiler will optimize this anyway)

Reply:
👍